module/zfs/dmu_send.c

```diff
@@ -2870,37 +2921,60 @@ dmu_adjust_send_estimate_for_indirects(dsl_dataset_t *ds, uint64_t uncompressed,
 }
 
 int
-dmu_send_estimate_fast(dsl_dataset_t *ds, dsl_dataset_t *fromds,
+dmu_send_estimate_fast(dsl_dataset_t *origds, dsl_dataset_t *fromds,
     zfs_bookmark_phys_t *frombook, boolean_t stream_compressed, uint64_t *sizep)
 {
 	int err;
+	dsl_dataset_t *ds = origds;
 	uint64_t uncomp, comp;
+	char dsname[ZFS_MAX_DATASET_NAME_LEN + 6];
 
-	ASSERT(dsl_pool_config_held(ds->ds_dir->dd_pool));
+	ASSERT(dsl_pool_config_held(origds->ds_dir->dd_pool));
 	ASSERT(fromds == NULL || frombook == NULL);
 
+	/*
+	 * If this is a saved send we may actually be sending
+	 * from the %recv clone used for resuming.
+	 */
+	if (!ds->ds_is_snapshot) {
```

Comment:
Should we be getting a flag to indicate that this is a saved send?  Otherwise, it seems like we could be taking on `%recv` when they are actually estimating a send to the filesystem itself (which I think is allowed as long as it isn't mounted writeable).

Reply:
This hasn't been allowed in the past. The old code just did:

```
	if (!ds->ds_is_snapshot)
		return (SET_ERROR(EINVAL));
```

Reply:
I spoke to Brian about this offline and we don't believe that this change should be needed. We came to the conclusion that the send estimate code shouldn't really have to care about whether this is a saved send or not. Let me know if you have any other thoughts.